File: packages/go-agentenkit/admin/default.go
// Package admin picks the operational store SetupAgentCore uses when none is
// configured (§2.9).
package admin

import (
	"context"
	"log/slog"
	"os"

	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/adapters/sqlite"
	postgresadmin "github.com/eadwinCode/agentic-kit/packages/go-agentenkit/admin/postgres"
	sqliteadmin "github.com/eadwinCode/agentic-kit/packages/go-agentenkit/admin/sqlite"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/ports"
)

// DefaultAdminDB is where the default store writes when no Postgres URL is set.
const DefaultAdminDB = "agentic-kit-admin.sqlite"

// OpenDefaultAdminStore is the store SetupAgentCore uses when none is
// configured (§2.9).
//
// AGENTIC_KIT_ADMIN_DATABASE_URL selects Postgres: point it at its own
// database or the one you already have, the agentic_ prefix keeps them
// apart. Without it, SQLite on disk at AGENTIC_KIT_ADMIN_DB.
//
// Like the TypeScript package, it uses whichever driver the process has:
// register a database/sql driver (modernc.org/sqlite or mattn/go-sqlite3;
// pgx/stdlib or lib/pq) by importing it.
//
// The connection is opened eagerly, so a store that cannot be opened fails at
// startup rather than losing every run record. The SCHEMA is brought up to
// date behind it (§2.9): a service starts at the same speed whether or not it
// has migrating to do, and the first admin call waits for the schema rather
// than meeting a table that is not there yet.
func OpenDefaultAdminStore(ctx context.Context, log *slog.Logger) (ports.AdminStore, error) {
	if url := os.Getenv("AGENTIC_KIT_ADMIN_DATABASE_URL"); url != "" {
		db, err := postgresadmin.Open(url)
		if err != nil {
			return nil, err
		}
		return postgresadmin.ConnectWithLogger(ctx, db, log)
	}
	file := os.Getenv("AGENTIC_KIT_ADMIN_DB")
	if file == "" {
		file = DefaultAdminDB
	}
	db, err := sqlite.Open(file)
	if err != nil {
		return nil, err
	}
	return sqliteadmin.NewContext(ctx, db, log)
}

File: packages/agentenkit/src/admin/default.ts
import type { AdminStore } from '../ports/admin.js';
import { openSqlite } from '../adapters/sqlite.js';
import { SqliteAdminStore } from './sqlite.js';
import { PostgresAdminStore, openPostgres } from './postgres.js';

/** Where the default store writes when no Postgres URL is set. */
export const DEFAULT_ADMIN_DB = 'agentic-kit-admin.sqlite';

/** The store `setupAgentCore` uses when none is configured (§2.9).
 *
 *  `AGENTIC_KIT_ADMIN_DATABASE_URL` selects Postgres — point it at its own
 *  database or the one you already have, the `agentic_` prefix keeps them
 *  apart. Without it, SQLite on disk at `AGENTIC_KIT_ADMIN_DB`, with whichever
 *  driver the process has.
 *
 *  The connection is opened eagerly. A store that cannot be opened is a
 *  configuration problem worth failing on at startup: degrading quietly would
 *  lose every run record, and a dashboard showing nothing looks the same as no
 *  traffic.
 *
 *  The SCHEMA is brought up to date behind it: a service starts at the same
 *  speed whether or not it has migrating to do, and the first admin call waits
 *  for the schema rather than meeting a table that is not there yet. */
export async function openDefaultAdminStore(
  log?: { error(message: string, ...rest: unknown[]): void },
): Promise<AdminStore> {
  const url = process.env.AGENTIC_KIT_ADMIN_DATABASE_URL;
  if (url) return PostgresAdminStore.connect(await openPostgres(url), log);
  return SqliteAdminStore.open(
    await openSqlite(process.env.AGENTIC_KIT_ADMIN_DB ?? DEFAULT_ADMIN_DB),
    log,
  );
}
